### PR TITLE
internal/worker/server: return an error on depsolve timeout (HMS-2989)

### PR DESCRIPTION
**Pull request by schuellerf**
Fixes the special case that if no worker is available and we generate an internal timeout and cancel the depsolve, including all followup jobs, no error was propagated.

For details, see https://issues.redhat.com/browse/HMS-2989

Now all jobs (in the DB) are set to "failed":
 * All `jobs` entries get a "job_error" in the `result` as well as
 * `started_at` and `failed_at` are set to "now" and
 * `token` is set to a random UUID to comply with the DB schema

This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
- [x] adequate documentation informing people about the change such as
  - [x] submit a PR for the READMEs [listed here](https://github.com/osbuild/osbuild.github.io/blob/main/readme-list)
  - [x] submit a PR for the [osbuild.org website](https://github.com/osbuild/osbuild.github.io/) repository if this PR changed any behavior not covered by the automatically updated READMEs

This is set to draft for now until:
* I get feedback that this approach is ok
* tests are implemented
* we decided on if we should "create a separate metrics" (see comment in the code)

### CODE REVIEW

**Comment by schuellerf** [comment]
On the frontend this looks like this:
Screenshot_20241004_120837

File: pkg/jobqueue/dbjobqueue/dbjobqueue.go [hunk 1]
```diff
@@ -94,6 +94,12 @@ const (
 		WHERE id = $1 AND finished_at IS NULL
 		RETURNING type, started_at`
 
+	sqlFailJob = `
```
[the remark this answers is not in the document]
Comment (croissanne) [reply pkg/jobqueue/dbjobqueue/dbjobqueue.go:97]: Actually this is fine to have a separate query.

[the remark this answers is not in the document]
Comment (schuellerf) [reply pkg/jobqueue/dbjobqueue/dbjobqueue.go:97]: From how I understood the behavior of the DB, the token, started_at, finished_at have to be set in order to get to a "failed" state where no jobs have been started (our case here)
otherwise it might appear in "requeue" and other queries

[the remark this answers is not in the document]
Comment (schuellerf) [reply pkg/jobqueue/dbjobqueue/dbjobqueue.go:97]: Hopefully I understood this correctly

File: internal/cloudapi/v2/server.go [hunk 2]
```diff
@@ -435,23 +436,80 @@ func (s *Server) enqueueKojiCompose(taskID uint64, server, name, version, releas
...
 	var dynArgs []json.RawMessage
 	var err error
+	token := uuid.Nil
 	logWithId := logrus.WithField("jobId", manifestJobID)
+
+	defer func() {
+		// token == uuid.Nil indicates that no worker even started processing
+		if token == uuid.Nil {
+			if jobResult.JobError != nil {
+				// set all jobs to "failed"
```
[the remark this answers is not in the document]
Comment (schuellerf) [reply internal/cloudapi/v2/server.go:461]: The error message in `result` column would be something else right? (not indicating the originating problem?)

**Comment by schuellerf** [comment]
@croissanne last question for me is:
should we create separate metrics? maybe in a followup PR?
```
func (s *Server) SetFailed(id uuid.UUID, error *clienterrors.Error) error {
	/* create a separate metrics?
…
```

File: internal/cloudapi/v2/server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (croissanne) [reply internal/cloudapi/v2/server.go:461]: Well that's why we chain the errors when building the error message. If done correctly the errors of all child jobs should show up, and in image builder we have some logic to surface nested errors based on error codes https://github.com/osbuild/image-builder/blob/de52a87ef277cc09d2728b49554d0da4166cc9a9/internal/v1/handler.go#L387-L419

So depending on the error codes set, the error we attach to the manifest job should surface.

[the remark this answers is not in the document]
Comment (croissanne) [reply internal/cloudapi/v2/server.go:461]: Also see https://github.com/osbuild/osbuild-composer/blob/6ccfc7f81884eedbf3290bb2118d0c6e31c676d5/internal/worker/server.go#L307-L320

**Comment by croissanne** [comment]
> @croissanne last question for me is: should we create separate metrics? maybe in a followup PR?
> 
> ```
> func (s *Server) SetFailed(id uuid.UUID, error *clienterrors.Error) error {
> 	/* create a separate metrics?
> …
> ```

I don't think that's necessary, the jobs will be marked as failed like any other job. As long as the job isn't marked as "userinputerror" it should trigger alerts and all that. Plus at that point presumable the depsolve job duration alert would also fire.

https://github.com/osbuild/osbuild-composer/blob/6ccfc7f81884eedbf3290bb2118d0c6e31c676d5/internal/worker/clienterrors/errors.go#L73-L111

File: Makefile [hunk 3]
```diff
@@ -304,6 +304,7 @@ container_composer_golangci_built.info: Makefile Containerfile_golangci_lint too
 
 .PHONY: lint
 lint: $(GOLANGCI_LINT_CACHE_DIR) container_composer_golangci_built.info
+	go fmt ./...
```
Comment (croissanne) [inline Makefile:307]: Might need a specific go version here, sometimes between major versions fmt works differently, and people might have different default go versions than the one used by `tools/prepare-source`

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -67,7 +67,7 @@ jobs:
       run: sudo apt-get update
       # gpgme-devel is needed for container upload dependencies
     - name: Install test dependencies
-      run: sudo apt-get install -y libgpgme-dev
+      run: sudo apt-get install -y libgpgme-dev libbtrfs-dev
```
Comment (croissanne) [inline .github/workflows/tests.yml:70]: Seems like an unrelated change? Could we split this out to a separate commit?

File: internal/cloudapi/v2/server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (croissanne) [reply internal/cloudapi/v2/server.go:461]: Hmm in that case I think it should be sufficient to set the resolve jobs and manifest job as failed, and let the build job take care of itself.

And then set the relevant timeout waiting for … errors on each of the resolve job, the manifest job will fail with a dependency error, and the osbuild will also fail with a dependency error. That way the chained errors will work correctly.

File: .github/workflows/tests.yml (hunk 4 above)
Comment (schuellerf) [reply .github/workflows/tests.yml:70]: can be even part of #4431

File: Makefile (hunk 3 above)
Comment (schuellerf) [reply Makefile:307]: Nice catch, thanks!

File: .github/workflows/tests.yml (hunk 4 above)
Comment (schuellerf) [reply .github/workflows/tests.yml]: The DB tests seem to fail if this is missing :-/
I'll add it for now (although I don't understand why this is needed now)

**Comment by schuellerf** [comment]
…rebased

File: internal/cloudapi/v2/server.go [hunk 5]
```diff
@@ -445,22 +446,79 @@ func (s *Server) enqueueKojiCompose(taskID uint64, server, name, version, releas
...
+				// osbuild job will fail as dependency
+				jobs := map[string]uuid.UUID{
```
Comment (croissanne) [inline internal/cloudapi/v2/server.go:472]: I wonder if it's ok to use a map here, since that means the loop below could be out of order. I'm not sure if there's a problem with failing the manifest job before the resolve jobs.

**Review by croissanne — COMMENTED** [review]
1 last comment, otherwise lgtm!

File: internal/cloudapi/v2/server.go [hunk 6]
```diff
@@ -445,22 +446,79 @@ func (s *Server) enqueueKojiCompose(taskID uint64, server, name, version, releas
...
 			select {
 			case <-ctx.Done():
-				logWithId.Warning("Manifest job dependencies took longer than 5 minutes to finish, or the server is shutting down, returning to avoid dangling routines")
+				logWithId.Warning(fmt.Sprintf("Manifest job dependencies took longer than %d minutes to finish,"+
+					" or the server is shutting down, returning to avoid dangling routines", depsolveTimeout))
+
+				jobResult.JobError = clienterrors.New(clienterrors.ErrorDepsolveTimeout,
+					"Timeout while waiting for package dependency resolution",
+					"There may be a temporary issue with compute resources. "+
+						"We’re looking into it, please try again later.",
```
Comment (croissanne) [inline internal/cloudapi/v2/server.go:520]: Not sure about the error message here, plus the alignment looks off to me? why is the `We're` more indented than the rest.

Also we don't tend to say that we're looking into it, people should be encouraged to open tickets when necessary. I'd just keep the first line really. Also technically it's not just package dependency, it's all resolve jobs.

We could ask the user to make sure all included sources are reachable and responsive etc…

Comment (schuellerf) [reply internal/cloudapi/v2/server.go]: I removed the last line for now, but ` please try again later.` is at least some hint what to do - I'm not sure if it's good to instruct them to contact support or leave them without a help "what to do"

**Review by croissanne — APPROVED** [review]
